File: c_impl/src/util/script_thread.c
#include "script_thread.h"
#include "global_allocator.h"
#include "panic.h"
#include <stdio.h>

#if defined(_WIN32) || defined(WIN32)

#define NOMINMAX
#define WIN32_LEAN_AND_MEAN
#include <windows.h>

typedef struct {
    HANDLE thread;
    DWORD identifier;
    /// May be NULL
    bool closeWithScript;
} CubsThreadWindowsImpl;

static DWORD WINAPI windows_thread_loop(CubsThreadWindowsImpl* self) {
    return 0;
}

static void windows_thread_close(CubsThreadWindowsImpl* self) {
    WaitForSingleObject(self->thread, INFINITE);
    CloseHandle(self->thread);
    cubs_free((void*)self, sizeof(CubsThreadWindowsImpl), _Alignof(CubsThreadWindowsImpl));
}

/// If `self` has an owner, don't bother freeing the thread on script close.
static void windows_thread_on_script_close(CubsThreadWindowsImpl* self) {
    if(self->closeWithScript == false) {
        return;
    }

    windows_thread_close(self);
}

static int windows_thread_get_id(const CubsThreadWindowsImpl* self) {
    return (int)self->identifier;
}

const CubsThreadVTable windowsVTable = {
    .onScriptClose = (CubsThreadOnScriptClose)&windows_thread_on_script_close, 
    .getId = (CubsThreadGetId)&windows_thread_get_id,
    .close = (CubsThreadClose)&windows_thread_close,
};

CubsThread cubs_thread_spawn(bool closeWithScript)
{
    CubsThread thread = {.threadObj = NULL, .vtable = NULL};

    CubsThreadWindowsImpl* impl = cubs_malloc(sizeof(CubsThreadWindowsImpl), _Alignof(CubsThreadWindowsImpl));

    DWORD identifier;
    HANDLE handle = CreateThread(
        NULL, // default security attributes
        0,  // default stack size
        (LPTHREAD_START_ROUTINE)windows_thread_loop, // function name
        impl, // argument to function
        0, // default creation flags
        &identifier); // thread identifier

    if(handle == NULL) {
        cubs_free((void*)impl, sizeof(CubsThreadWindowsImpl), _Alignof(CubsThreadWindowsImpl));
        cubs_panic("Failed to spawn CubicScript thread");
    }

    impl->thread = handle;
    impl->identifier = identifier;
    impl->closeWithScript = closeWithScript;

    thread.threadObj = (void*)impl;
    thread.vtable = &windowsVTable;

    
    return thread;
}

#endif // WIN32

int cubs_thread_get_id(const CubsThread *thread)
{
    return thread->vtable->getId(thread->threadObj);
}

void cubs_thread_close(CubsThread *thread)
{
    if(thread->vtable->close == NULL) {
        return;
    }
    thread->vtable->close(thread->threadObj);
}

File: c_impl/src/util/script_thread.zig
const std = @import("std");
const expect = std.testing.expect;

const c = struct {
    extern fn cubs_thread_spawn(closeWithScript: bool) callconv(.C) ScriptThread;
    extern fn cubs_thread_close(thread: *ScriptThread) callconv(.C) void;
    extern fn cubs_thread_get_id(thread: *const ScriptThread) callconv(.C) u32;
};

pub const ScriptThread = extern struct {
    const Self = @This();

    threadObj: *anyopaque,
    vtable: *const VTable,

    /// Create an instance of an OS specific script thread. Implemention is defined in `script_thread.c`.
    pub fn spawn() Self {
        return c.cubs_thread_spawn(false);
    }

    /// Explicitly close the thread. If `self.vtable.close == null`, does nothing.
    pub fn close(self: *Self) void {
        c.cubs_thread_close(self);
    }

    pub fn getId(self: *const Self) u32 {
        return c.cubs_thread_get_id(self);
    }

    pub const VTable = extern struct {
        onScriptClose: ?*const fn (threadObj: *anyopaque) callconv(.C) void = null,
        getId: *const fn (threadObj: *const anyopaque) callconv(.C) c_int,
        close: ?*const fn (threadObj: *anyopaque) callconv(.C) void = null,
    };

    test spawn { // also tests close
        var thread = Self.spawn();
        defer thread.close();
    }

    test getId {
        var thread = Self.spawn();
        defer thread.close();

        try expect(@as(u32, @intCast(std.Thread.getCurrentId())) != thread.getId());
    }
};
